### DISCUSSION TITLE
Why hasn't Docker Rootless implemented cgroup v1? Are there any limitations?

### DISCUSSION THREAD

**Discussion by yushiqie** [moby/moby#48778] [Q&A] [answered]
Why hasn't Docker Rootless implemented cgroup v1? Are there any limitations?

**Comment by AkihiroSuda** [comment, marked as answer]
Because there was no easy and secure way to delegate cgroup v1 to non-root users.
You can still run Rootless Docker on cgroup v1 hosts, but cgroup-related flags such as `--cpus` and `--memory` are just ignored.

Most Linux distributions have already switched to cgroup v2 that properly supports rootless delegation.
https://rootlesscontaine.rs/getting-started/common/cgroup2/

**Reply by yushiqie** [reply]
I'm a bit confused about this: the root user can support both cgroupv1 and cgroup v2, while non-root users do not support cgroupv1. Is this more of a security consideration or an implementation cost issue? Because from an implementation perspective, if cgroupv1 itself were readable and writable, it seems like it could support cgroupv1. So, is it more of a security concern?
